### PR TITLE
[PyTorch] Fixes and tests for FP8 + activation recompute

### LINKED ISSUE #448: Recomputation makes flash-attn backward slow
Hi, it seems that an entire recomputation of flash-attn makes the backward stage quite slow while `checkpoint_core_attention` is triggered. The overhead can be several `contiguous` ops and a `flash_attn_fwd` op. 

One potential scenario arises when training with Megatron-LM and enabling the `--recompute-activation` flag. This overhead of the `transformer_engine` impl becomes evident since Megatron-LM doesn't use checkpoint for flash-attn.

### PR DESCRIPTION
**Pull request by ksivaman**
Fixes #448

### CODE REVIEW

File: transformer_engine/pytorch/module/base.py [hunk 1]
```diff
@@ -588,7 +592,6 @@ def prepare_forward(
                 self.fp8
                 and self.training
                 and is_fp8_activation_recompute_enabled()
-                and not in_fp8_activation_recompute_phase()
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/base.py:591]: Do we expect this change to affect behavior? I can imagine some pathological cases where we do multiple recomputations in the backward pass, e.g. recursive activation checkpointing.

**Review by timmoon10 — APPROVED** [review]
Overall LGTM, pending CI.

File: transformer_engine/pytorch/module/base.py (hunk 1 above)
Comment (ksivaman) [reply transformer_engine/pytorch/module/base.py:591]: Because of the logic here, this condition is always `True`.

**Pull request merged**